### PR TITLE
UAR-616: Overseas entity email displayed to user

### PR DESCRIPTION
**Pull request by conuirchgov**
### JIRA link

https://companieshouse.atlassian.net/browse/UAR-616

### Change description

- Added API call to retrieve Private OE details
- Populated email field on OE Query page

### Associated PRs:
- https://github.com/companieshouse/overseas-entities-api/pull/277
- https://github.com/companieshouse/api-sdk-node/pull/571

TODO: Should a failure to retrieve private details result in a Service Unavailable error or a Company Not Found validation error (which is already there for a failure to get `companyProfile` info)?

### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/controllers/update/overseas.entity.query.controller.ts [hunk 1]
```diff
@@ -32,7 +33,9 @@ export const post = async (req: Request, res: Response, next: NextFunction) => {
 
     const entityNumber = req.body[EntityNumberKey];
     const companyProfile = await getCompanyProfile(req, entityNumber);
-    if (!companyProfile) {
+    const privateOeDetails = await getPrivateOeDetails(req, entityNumber);
+    // TO-DO: Need to handle error for no email
```
Comment (conuirchgov) [inline src/controllers/update/overseas.entity.query.controller.ts:37]: Should a failure to retrieve private details result in a Service Unavailable error or a Company Not Found validation error (which is already there for a failure to get companyProfile info)?

File: src/controllers/update/overseas.entity.query.controller.ts [hunk 2]
```diff
@@ -32,7 +33,9 @@ export const post = async (req: Request, res: Response, next: NextFunction) => {
 
     const entityNumber = req.body[EntityNumberKey];
     const companyProfile = await getCompanyProfile(req, entityNumber);
-    if (!companyProfile) {
+    const privateOeDetails = await getPrivateOeDetails(req, entityNumber);
```
Comment (AndreKramer7654) [inline src/controllers/update/overseas.entity.query.controller.ts:36]: Only fetch if companyProfile is not already fetched.

File: src/controllers/update/overseas.entity.query.controller.ts (hunk 1 above)
Comment (AndreKramer7654) [reply src/controllers/update/overseas.entity.query.controller.ts:37]: Treat same as company profile error.

File: src/service/private.overseas.entity.details.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  const response = await makeApiCallWithRetry(
+    "overseasEntity",
+    "getOverseasEntityDetails",
+    req,
+    req.session as Session,
+    oeNumber,
+  );
+
+  if (response.httpStatusCode !== 200 && response.httpStatusCode !== 404) {
+    const errorMsg = `Something went wrong fetching company details = ${JSON.stringify(response)}`;
```
Comment (AndreKramer7654) [inline src/service/private.overseas.entity.details.ts:20]: Do not print out the whole response. Just the status code & any error message.

File: src/controllers/update/overseas.entity.query.controller.ts (hunk 2 above)
Comment (AndreKramer7654) [reply src/controllers/update/overseas.entity.query.controller.ts:36]: Make sure you fetch the company profile first and handle any errors. Only then attempt the email fetch.

File: src/service/private.overseas.entity.details.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    throw createAndLogErrorRequest(req, errorMsg);
+  }
+
+  if (response.httpStatusCode === 404) {
+    logger.debugRequest(req, `No company data found for ${oeNumber}`);
```
Comment (AndreKramer7654) [inline src/service/private.overseas.entity.details.ts:25]: This deserves a error log message too.

File: src/service/private.overseas.entity.details.ts (hunk 3 above)
Comment (AndreKramer7654) [reply src/service/private.overseas.entity.details.ts:20]: Log error for failure before throw.

Comment (conuirchgov) [reply src/service/private.overseas.entity.details.ts:20]: Why would we not want to log the `response` specifically in this instance?

And on line 21, the error is already being logged before it's thrown 👍

File: src/service/private.overseas.entity.details.ts (hunk 4 above)
Comment (conuirchgov) [reply src/service/private.overseas.entity.details.ts:25]: Similar to the response to the comment on line 20, this is the same as the logging for fetching the company profile data. Any particular reason why they should differ or should company profile also be updated?

File: src/controllers/update/overseas.entity.query.controller.ts (hunk 2 above)
Comment (conuirchgov) [reply src/controllers/update/overseas.entity.query.controller.ts]: Now calling `getPrivateOeDetails` only if `companyProfile` exists.

**Review by stephanie-ch — APPROVED** [review]

File: src/utils/update/company.profile.mapper.to.overseas.entity.ts [hunk 5]
```diff
@@ -13,7 +14,7 @@ export const mapCompanyProfileToOverseasEntity = (cp: CompanyProfile): Entity =>
     incorporation_country: cp.jurisdiction,
     public_register_name: cp.foreignCompanyDetails?.originatingRegistry?.name,
     public_register_jurisdiction: cp.foreignCompanyDetails?.originatingRegistry?.country,
-    email: "", // private data
+    email: oEDetails.email_address, // private data
```
[the remark this answers is not in the document]
Comment (conuirchgov) [reply src/utils/update/company.profile.mapper.to.overseas.entity.ts]: Updated 👍

File: src/service/private.overseas.entity.details.ts (hunk 3 above)
Comment (AndreKramer7654) [reply src/service/private.overseas.entity.details.ts:20]: error code would be enough here. Don't stringify JSON

File: src/service/private.overseas.entity.details.ts (hunk 4 above)
Comment (AndreKramer7654) [reply src/service/private.overseas.entity.details.ts:25]: not too concerned about this one to be fair.

File: src/controllers/update/overseas.entity.query.controller.ts [hunk 6]
```diff
@@ -30,31 +32,41 @@ export const post = async (req: Request, res: Response, next: NextFunction) => {
   try {
     logger.debugRequest(req, `${req.method} ${req.route.path}`);
 
+    let privateOeDetails: OverseasEntityExtraDetails | undefined;
+
     const entityNumber = req.body[EntityNumberKey];
     const companyProfile = await getCompanyProfile(req, entityNumber);
-    if (!companyProfile) {
+
+    if (companyProfile) {
```
[the remark this answers is not in the document]
Comment (conuirchgov) [reply src/controllers/update/overseas.entity.query.controller.ts]: We decided to move it to as early in the journey as possible, so it's been moved to the `GET` int the `update filing date` controller, which comes directly after OAuth.

File: src/service/private.overseas.entity.details.ts (hunk 3 above)
Comment (conuirchgov) [reply src/service/private.overseas.entity.details.ts:20]: Going to leave it as is for consistency with the fetch in `company.profile.service.ts`.

File: src/utils/update/company.profile.mapper.to.overseas.entity.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (conuirchgov) [reply src/utils/update/company.profile.mapper.to.overseas.entity.ts]: No longer relevant as the private details are now added to the session data at a different stage of the journey.

File: src/controllers/update/update.filing.date.controller.ts [hunk 7]
```diff
@@ -1,22 +1,42 @@
...
 import { ApplicationData } from "../../model/application.model";
+import { getPrivateOeDetails } from "../../service/private.overseas.entity.details";
 
-export const get = (req: Request, res: Response, next: NextFunction) => {
+const fetchAndSavePrivateOeData = async (req: Request, appData: ApplicationData): Promise<void> => {
+  const companyNumber = appData[EntityNumberKey];
+
+  if (!companyNumber) {
+    throw createAndLogErrorRequest(req, "Unable to request private OE data. Missing entity number in application data");
+  }
```
Comment (corybudd-ch) [inline src/controllers/update/update.filing.date.controller.ts:20]: this route has a navcheck for `hasOverseasEntity` so I don't think this code of !companyNumber will ever be a genuine scenario and if it's needed or not.

**Review by corybudd-ch — APPROVED** [review]

**Review by stephanie-ch — APPROVED** [review]

File: src/controllers/update/update.filing.date.controller.ts [hunk 8]
```diff
@@ -1,22 +1,42 @@
...
+};
+
+export const get = async(req: Request, res: Response, next: NextFunction) => {
   try {
     logger.debugRequest(req, `${req.method} ${req.route.path}`);
+
     const appData = getApplicationData(req.session);
     const filingDate = appData.update?.[FilingDateKey] ? mapDataObjectToFields(appData.update[FilingDateKey], FilingDateKeys, InputDateKeys) : {};
 
+    await fetchAndSavePrivateOeData(req, appData);
```
Comment (AndreKramer7654) [inline src/controllers/update/update.filing.date.controller.ts:38]: If the user edits the email address but goes back to this page the email will be fetched again. Overwriting the user edit. Email addr needs to be store in OE entity part of model and there should be a check to see if it is already fetched (not null or empty in OE entity) and only fetch if not.

Comment (conuirchgov) [reply src/controllers/update/update.filing.date.controller.ts]: Good catch :thumbsup: We now only fetch if it hasn't been already.

**Review by AndreKramer7654 — APPROVED** [review]

**Comment by conuirchgov** [comment]
We've agreed to go ahead with merging despite quality gate failure due to duplicate code (linked below).

The duplicate code is a service file written for making API calls, however the pattern in place is to have a separate file per endpoint so this can't be avoided.

[SonarQube failure: Code Duplication exceeds 3%](https://code-analysis.platform.aws.chdev.org/component_measures?id=uk.gov.companieshouse%3Aoverseas-entities-web&metric=new_duplicated_lines_density&pullRequest=916&view=list)

**Pull request merged**